File: lib/biz/spec-iot/src/test/java/site/addzero/biz/spec/iot/tdengine/TdSchemaPlannerTest.java
package site.addzero.biz.spec.iot.tdengine;

import org.junit.jupiter.api.Test;

import java.util.Arrays;
import java.util.Collections;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

class TdSchemaPlannerTest {

    @Test
    void shouldPlanAddDropModifyAndRecreateColumns() {
        TdSchemaPlanner planner = new TdSchemaPlanner();
        TdSchemaDiff diff = planner.plan(
                Arrays.asList(
                        TdengineSchemaDefaults.tsColumn(),
                        TdengineSchemaDefaults.reportTimeColumn(),
                        TdengineSchemaDefaults.deviceIdTagColumn(),
                        new TdColumnSpec("temperature", TdColumnSpec.TYPE_INT, null, null),
                        new TdColumnSpec("legacy", TdColumnSpec.TYPE_FLOAT, null, null),
                        new TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 64, null),
                        new TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 32, null)
                ),
                Arrays.asList(
                        new TdColumnSpec("temperature", TdColumnSpec.TYPE_FLOAT, null, null),
                        new TdColumnSpec("status", TdColumnSpec.TYPE_BOOL, null, null),
                        new TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 128, null),
                        new TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 16, null)
                )
        );

        assertEquals(Collections.singletonList(new TdColumnSpec("status", TdColumnSpec.TYPE_BOOL, null, null)), diff.getAddedColumns());
        assertEquals(Collections.singletonList(new TdColumnSpec("legacy", TdColumnSpec.TYPE_FLOAT, null, null)), diff.getDroppedColumns());
        assertEquals(Collections.singletonList(new TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 128, null)), diff.getModifiedColumns());
        assertEquals(
                Arrays.asList(
                        new TdColumnSpec("temperature", TdColumnSpec.TYPE_FLOAT, null, null),
                        new TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 16, null)
                ),
                diff.getRecreatedColumns()
        );
    }

    @Test
    void shouldReturnEmptyDiffForEmptySchema() {
        TdSchemaPlanner planner = new TdSchemaPlanner();
        TdSchemaDiff diff = planner.plan(Collections.<TdColumnSpec>emptyList(), Collections.<TdColumnSpec>emptyList());
        assertTrue(diff.isEmpty());
    }
}

File: lib/biz/spec-iot/src/test/kotlin/site/addzero/biz/spec/iot/tdengine/TdSchemaPlannerTest.kt
package site.addzero.biz.spec.iot.tdengine

import org.junit.jupiter.api.Test
import kotlin.test.assertEquals
import kotlin.test.assertTrue

class TdSchemaPlannerTest {

    @Test
    fun shouldPlanAddDropModifyAndRecreateColumns() {
        val planner = TdSchemaPlanner()
        val diff = planner.plan(
            listOf(
                TdengineSchemaDefaults.tsColumn(),
                TdengineSchemaDefaults.reportTimeColumn(),
                TdengineSchemaDefaults.deviceIdTagColumn(),
                TdColumnSpec("temperature", TdColumnSpec.TYPE_INT, null, null),
                TdColumnSpec("legacy", TdColumnSpec.TYPE_FLOAT, null, null),
                TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 64, null),
                TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 32, null),
            ),
            listOf(
                TdColumnSpec("temperature", TdColumnSpec.TYPE_FLOAT, null, null),
                TdColumnSpec("status", TdColumnSpec.TYPE_BOOL, null, null),
                TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 128, null),
                TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 16, null),
            ),
        )

        assertEquals(
            listOf(TdColumnSpec("status", TdColumnSpec.TYPE_BOOL, null, null)),
            diff.addedColumns,
        )
        assertEquals(
            listOf(TdColumnSpec("legacy", TdColumnSpec.TYPE_FLOAT, null, null)),
            diff.droppedColumns,
        )
        assertEquals(
            listOf(TdColumnSpec("name", TdColumnSpec.TYPE_VARCHAR, 128, null)),
            diff.modifiedColumns,
        )
        assertEquals(
            listOf(
                TdColumnSpec("temperature", TdColumnSpec.TYPE_FLOAT, null, null),
                TdColumnSpec("mode", TdColumnSpec.TYPE_VARCHAR, 16, null),
            ),
            diff.recreatedColumns,
        )
    }

    @Test
    fun shouldReturnEmptyDiffForEmptySchema() {
        val planner = TdSchemaPlanner()
        val diff = planner.plan(emptyList(), emptyList())
        assertTrue(diff.isEmpty())
    }
}
